opm/simulators/wells/BlackoilWellModelConstraints.cpp

```diff
@@ -0,0 +1,479 @@
+            current_rate += WellGroupHelpers::sumWellResRates(group, well_.schedule(), well_state, reportStepIdx, pu.phase_pos[BlackoilPhases::Vapour], true);
+
+            // sum over all nodes
+            current_rate = well_.comm().sum(current_rate);
+
+            double target = controls.resv_target;
+            if (group.has_gpmaint_control(Group::ProductionCMode::RESV))
+                target = well_.groupState().gpmaint_target(group.name());
+
+            if ( target < current_rate ) {
+                double scale = 1.0;
+                if (current_rate > 1e-12)
+                    scale = target / current_rate;
+                return std::make_pair(Group::ProductionCMode::RESV, scale);
+            }
+        }
+    }
+    if (group.has_control(Group::ProductionCMode::PRBL))
+    {
+        OPM_DEFLOG_THROW(std::runtime_error, "Group " + group.name() + "PRBL control for production groups not implemented", deferred_logger);
+    }
+    return std::make_pair(Group::ProductionCMode::NONE, 1.0);
+}
+
+bool BlackoilWellModelConstraints::
+checkGroupConstraints(const Group& group,
+                      const int reportStepIdx,
+                      DeferredLogger& deferred_logger) const
+{
+    if (group.isInjectionGroup()) {
+        const Phase all[] = {Phase::WATER, Phase::OIL, Phase::GAS};
+        for (Phase phase : all) {
+            if (!group.hasInjectionControl(phase)) {
+                continue;
+            }
+            const auto& check = this->checkGroupInjectionConstraints(group,
+                                                                     reportStepIdx,  phase);
+            if (check.first != Group::InjectionCMode::NONE) {
+                return true;
+            }
+        }
+    }
+    if (group.isProductionGroup()) {
+        const auto& check = this->checkGroupProductionConstraints(group,
+                                                                  reportStepIdx,
+                                                                  deferred_logger);
+        if (check.first != Group::ProductionCMode::NONE)
+        {
+            return true;
+        }
+    }
+
+    // call recursively down the group hiearchy
+    bool violated = false;
+    for (const std::string& groupName : group.groups()) {
+        const auto& grp = well_.schedule().getGroup(groupName, reportStepIdx);
+        violated = violated || checkGroupConstraints(grp, reportStepIdx, deferred_logger);
+    }
+    return violated;
+}
+
+void BlackoilWellModelConstraints::
+actionOnBrokenConstraints(const Group& group,
+                          const Group::ExceedAction& exceed_action,
+                          const Group::ProductionCMode& newControl,
+                          GroupState& group_state,
+                          DeferredLogger& deferred_logger) const
+{
+    const Group::ProductionCMode oldControl = well_.groupState().production_control(group.name());
+
+    std::ostringstream ss;
```

Comment:
We should include `<sstream>` to bring a declaration of `std::ostringstream` into scope.  We might also consider switching this to using libfmt instead, but that's probably better suited to a follow-up PR.

Reply:
done.